File: vcdo/sources/gmail.py
"""Gmail: message metadata, and the attachment-bytes worker.

Two outputs from one fetch. Messages become records in the raw lake; attachments
become *objects* in the raw lake. They are separate datasets with separate
prefixes, and a successful message sync does **not** mean attachment ingestion
succeeded — the handoff is explicit, and it is the failure that matters most,
because message counts look healthy while documents are silently missing.

**Attachment identity never uses Gmail's ``attachmentId``.** Gmail regenerates it
on every fetch: the legacy system measured 0 of 114 matching across two identical
runs. An id that changes cannot be an idempotency key, cannot dedupe, and cannot
be dereferenced later. Identity here is ``(tenant_id, message_id, part_id)`` plus
the SHA-256 of the bytes — all derivable without any server-assigned value.

Identity used to lead with the *mailbox* and end with the *filename*. Both were
removed from the key: an address and an attachment name are personal data, and an
object key is the one part of the lake that appears in listings, logs and error
messages. Neither carried uniqueness the tenant and positional part id do not
already carry. Both are still recorded in the manifest.

That is why live mode fetches ``format="raw"``/``full`` once per message and
extracts bytes locally rather than making a second ``attachments.get`` call. One
read per message, and the frozen payload is the evidence for both datasets.
"""

from __future__ import annotations

import json
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path

from vcdo.sources.base import RawRecord, SourceError
from vcdo.sources.mime import attachment_parts, is_pdf, looks_truncated, sniff

__all__ = ["GmailSource", "Attachment", "AttachmentOutcome", "ENTITIES"]

ENTITIES = ("messages",)


@dataclass(frozen=True, slots=True)
class Attachment:
    tenant_id: str
    mailbox: str
    message_id: str
    part_id: str
    filename: str | None
    declared_mime: str | None
    data: bytes

    def lake_key(self) -> str:
        """Positional and stable. Never contains a server-assigned id.

        **Keyed by tenant, not by mailbox.** The mailbox is a person's email
        address, and an object key is the one part of the lake that surfaces in
        listings, logs and error messages; the tenant CASE-ID says the same thing
        about ownership without naming anyone. The address itself is recorded in
        the manifest, where access is controlled. There is exactly one Gmail
        connection per tenant (``ops.connection`` is keyed ``(tenant_id,
        source)``), so this is no less unique than the address was.

        **The filename is not in the key either.** It carried no uniqueness --
        ``part_id`` is positional and already unique within a message -- and
        attachment filenames routinely contain client names.
        """
        return f"gmail/{self.tenant_id}/attachments/{self.message_id}/{self.part_id}"


@dataclass(frozen=True, slots=True)
class AttachmentOutcome:
    """What happened to one candidate part.

    ``status`` is one of ``stored``, ``skipped_not_pdf``, or ``failed_corrupt``.
    Three states, not two: a corrupt file is neither successfully stored nor
    legitimately skipped, and collapsing it into either one hides a real problem.
    """

    status: str
    reason: str
    attachment: Attachment | None = None


class GmailSource:
    name = "gmail"

    def __init__(
        self,
        *,
        tenant_id: str,
        mode: str = "mock",
        fixtures_dir: str = "fixtures",
        credentials=None,
        mailbox: str = "",
    ) -> None:
        self.tenant_id = tenant_id
        #: The mailbox address this connection reads. Public, and deliberately
        #: separate from ``tenant_id``: one names the customer whose lake this
        #: is, the other names a person. They were the same field, which put an
        #: email address into every attachment's object key.
        self.mailbox = mailbox
        self.mode = mode
        self.fixtures_dir = Path(fixtures_dir)
        self._credentials = credentials
        if mode == "live" and credentials is None:
            raise SourceError("gmail live mode requires OAuth credentials")
        if mode == "live" and not mailbox:
            # Live reads use userId="me", so the mailbox is whatever the
            # credential happens to be. Recording which one we believe it is
            # makes a mis-connected account visible instead of silent.
            raise SourceError("gmail live mode requires mailbox (the address the credential reads)")

    def entities(self) -> tuple[str, ...]:
        return ENTITIES

    def read(self, entity: str) -> Iterator[RawRecord]:
        if entity not in ENTITIES:
            raise SourceError(f"unknown gmail entity {entity!r}")
        for message in self._messages():
            yield RawRecord(
                source=self.name,
                tenant_id=self.tenant_id,
                entity="messages",
                source_record_id=str(message.get("id") or ""),
                # internalDate is Gmail's own clock in epoch milliseconds. The
                # `Date:` header is the SENDER's clock and is routinely wrong or
                # forged, so ordering and windows must never use it.
                source_updated_at=_epoch_ms_to_iso(message.get("internalDate")),
                payload=message,
            )

    def _messages(self) -> list[dict]:
        if self.mode == "mock":
            path = self.fixtures_dir / "gmail" / "messages.json"
            if not path.exists():
                raise SourceError(f"no gmail fixture at {path}")
            data = json.loads(path.read_text())
            messages = data.get("messages")
            if not isinstance(messages, list):
                raise SourceError(f"fixture {path} has no 'messages' list")
            return messages
        return list(self._live_messages())

    def _live_messages(self) -> Iterator[dict]:
        """Fetch each message once, in full. No second attachments.get call."""
        from googleapiclient.discovery import build

        service = build("gmail", "v1", credentials=self._credentials, cache_discovery=False)
        try:
            page_token = None
            while True:
                listing = (
                    service.users()
                    .messages()
                    .list(userId="me", maxResults=500, pageToken=page_token)
                    .execute()
                )
                for stub in listing.get("messages", []):
                    yield (
                        service.users().messages().get(userId="me", id=stub["id"], format="full").execute()
                    )
                page_token = listing.get("nextPageToken")
                if not page_token:
                    break
        except Exception as exc:
            raise SourceError(f"gmail read failed: {exc}") from exc

    # -- the attachment worker ------------------------------------------------

    def extract_attachments(self, message: dict) -> list[AttachmentOutcome]:
        """Pull PDF bytes out of an already-fetched message.

        Pure with respect to the network in mock mode and in live mode alike:
        the bytes arrived with the message. That is what lets attachments be
        rebuilt from the frozen lake rather than re-fetched.
        """
        message_id = str(message.get("id") or "")
        outcomes: list[AttachmentOutcome] = []

        for part in attachment_parts(message.get("payload") or {}):
            if part.data is None:
                # Gmail supplied an id instead of inline bytes. We do not chase
                # it: the id is already stale. Recorded as a failure so it is
                # visible rather than silently absent.
                outcomes.append(
                    AttachmentOutcome(
                        "failed_corrupt",
                        f"part {part.part_id} carried no inline data "
                        f"(attachmentId is not usable as an identity)",
                    )
                )
                continue

            if looks_truncated(part.data):
                outcomes.append(
                    AttachmentOutcome(
                        "failed_corrupt",
                        f"{part.filename!r} begins with %PDF- but has no end marker",
                    )
                )
                continue

            if not is_pdf(part.data, declared_mime=part.mime_type, filename=part.filename):
                outcomes.append(
                    AttachmentOutcome(
                        "skipped_not_pdf",
                        f"{part.filename!r} is {sniff(part.data) or 'unrecognised'}, "
                        f"declared {part.mime_type!r}",
                    )
                )
                continue

            outcomes.append(
                AttachmentOutcome(
                    "stored",
                    "pdf confirmed by content",
                    Attachment(
                        tenant_id=self.tenant_id,
                        mailbox=self.mailbox,
                        message_id=message_id,
                        part_id=part.part_id,
                        filename=part.filename,
                        declared_mime=part.mime_type,
                        data=part.data,
                    ),
                )
            )
        return outcomes


def _epoch_ms_to_iso(value: object) -> str | None:
    from datetime import UTC, datetime

    if value in (None, ""):
        return None
    try:
        return datetime.fromtimestamp(int(value) / 1000, tz=UTC).isoformat()
    except (TypeError, ValueError):
        return None

File: apps/worker/src/services/google/gmail.ts
/**
 * Gmail: message headers into `raw.records`, matching attachments into the lake.
 *
 * What is promised on the consent card is what this reads and no more: "message headers and
 * the attachment types you allow, from the mailbox you connect", scoped to chosen labels or
 * deliberately to the whole mailbox. Bodies are never fetched.
 *
 * FOUR THINGS HERE ARE EASY TO GET WRONG AND EXPENSIVE TO GET WRONG.
 *
 * **`format` decides whether attachments exist at all.** `metadata` returns "only email
 * message ID, labels, and email headers" -- Google's own words -- which means no
 * `payload.parts` and so no `body.attachmentId`. A collector that asks for it walks an empty
 * parts list and lands every message with none of its attachments, looking for all the world
 * like a mailbox that has none. Only `format=full` carries them. This module asked for
 * `metadata` until the fix that added this paragraph, and the suite agreed with it because
 * the fixture answered a metadata request with parts -- an API that does not exist.
 *
 * **`labelIds` is AND, not OR.** One query carrying three label ids returns only the
 * messages that hold all three, which for most selections is none. A single query looks
 * right, passes a test written from the same misunderstanding, and silently ingests nothing
 * for every tenant that picked more than one label. So there is one paged query per selected
 * label, unioned by message id.
 *
 * **`attachmentId` is not stable.** It is scoped to one message read and changes between
 * fetches, so keying a document on it re-lands the same attachment under a new key every run
 * -- unbounded storage growth that looks like legitimate history. The stable identity is
 * `(messageId, partIndex)`.
 *
 * **Which types are allowed is `scope.fileTypes`, checked client-side.** Gmail's API has no
 * request-level filter on an attachment's MIME type -- only `labelIds` narrows what is fetched
 * at all -- so every part of every fetched message is walked and matched here. Empty means
 * every type, the same recorded-decision idiom `@undercroft/contracts` uses for an empty
 * label or entity list.
 *
 * Headers are extracted rather than stored whole, and that is now the ONLY thing keeping a
 * body out of the lake: `format=full` returns bodies and every header, where `format=metadata`
 * returned neither. Nothing here reads `body.data` or `snippet`, `headerMap` keeps the six
 * headers the consent names, and `messageRecord` builds its payload from extracted fields
 * rather than the response -- so what is fetched is wider than before while what is STORED is
 * byte-for-byte what it was. Both halves are pinned by tests; neither is safe to relax.
 */

import { allowsFile, type GmailScope, landedType } from "@undercroft/contracts";
import { canonicalJson, decodeBase64Url, getPath, getStringPath } from "@undercroft/core";

import type { DocumentToLand } from "../landDocument.ts";
import type { RecordToLand } from "../land.ts";
import type { RunJournal } from "../runJournal.ts";
import type { GoogleApi } from "./api.ts";
import type { AlreadyHeld, Harvest } from "./harvest.ts";

const GMAIL_BASE = "https://gmail.googleapis.com/gmail/v1/users/me";
const ENTITY = "messages";
/** Page size. Gmail caps at 500; 100 keeps a page's follow-up fetches bounded. */
const PAGE_SIZE = "100";
/** Enough to identify and reconcile a message; deliberately not the body. */
const HEADERS = ["From", "To", "Cc", "Subject", "Date", "Message-ID"] as const;

export function gmailBaseUrl(): string {
  return GMAIL_BASE;
}

/** One row of the label listing, as the scope picker needs to see it. */
export interface GmailLabel {
  readonly id: string;
  readonly name: string;
  /** Gmail's own `type`, or null where it did not say. See `labelKind`. */
  readonly kind: "system" | "user" | null;
}

/** The labels an admin can choose from. Needs a live token, so it runs in the worker. */
export async function listLabels(api: GoogleApi): Promise<GmailLabel[]> {
  const body = await api.getJson(`${GMAIL_BASE}/labels`, "labels");
  const labels = asArray(getPath(body, "labels"));
  return labels.flatMap((label) => {
    const id = str(label, "id");
    if (id === "") {
      return [];
    }
    const name = str(label, "name");
    return [{ id, name: name === "" ? id : name, kind: labelKind(str(label, "type")) }];
  });
}

/**
 * Gmail's ownership of a label, kept three-valued.
 *
 * The picker sets an admin's own labels ahead of the thirteen Gmail ships, because in a real
 * mailbox the built-ins are the noise that the labels they came to find are buried under.
 * That ordering is only worth having if it is TRUE, so a label Gmail did not classify stays
 * unclassified rather than being filed under the commoner of the two answers.
 */
function labelKind(reported: string): "system" | "user" | null {
  return reported === "system" || reported === "user" ? reported : null;
}

/**
 * Harvest a mailbox under a chosen scope, one message at a time.
 *
 * Yields what to land rather than landing it, so the decision of what a mailbox contains is
 * testable without a lake or a database -- and yields rather than returns, so the decision
 * costs one message rather than the whole mailbox. Buffering it was what put 7,786 messages
 * in a 1 GiB container and got the process oom-killed 76 minutes in.
 *
 * **A MESSAGE WE ALREADY HOLD, COMPLETE, IS NOT FETCHED AT ALL.** Listing ids is one request
 * per hundred; reading a message is one paced request each, at 334ms, so a 7,786-message
 * mailbox is 43 minutes of `messages.get` and nothing else. The test is a row in
 * `raw.records` that is MARKED harvest-complete -- a message whose LABELS changed since is
 * still skipped, which is a decision the user took with its cost stated: a second run that
 * takes a minute instead of forty-three, at the price of a relabelling we will not notice
 * until something else makes us read the message.
 *
 * PRESENCE ALONE WAS THE TEST FOR ONE RELEASE, AND IT LOST A MAILBOX. It rested on a message's
 * record not reaching `raw.records` until its attachments had reached the lake -- true of
 * every row that rule wrote, and untrue of every row written before it, which is all of them
 * on a tenant that had already run. The 7,786 above is not an example: it is the mailbox, with
 * 7,786 records and zero documents, skipped whole on every run until ADR 0035 made the mark
 * something a landing states rather than something this file asserts. `harvest.ts` records the
 * contract, `collect.ts` counts what got down, and `knownRecords` reads the count.
 *
 * **One consequence, so nobody reads the filter below as unconditional:** the defence-in-
 * depth label check now runs only on a message this run actually fetched. A message we
 * already hold is never fetched, so a label removed from it after it landed does not
 * un-land it. Nothing widens -- what was stored was covered by the selection when it was
 * stored -- but the check is a check on new reads, not a sweep over the mailbox.
 */
export async function* harvestGmail(
  api: GoogleApi,
  scope: GmailScope,
  journal: RunJournal,
  held: AlreadyHeld,
): Harvest {
  const selected = new Set(scope.labels.map((l) => l.id));
  const messageIds = await listMessageIds(api, scope);
  // No timestamp: the listing carries none, so the only question this probe can put is
  // whether we hold the message, complete. See `RecordProbe` and `knownRecords`.
  const known = await held(messageIds.map((id) => ({ sourceRecordId: id, sourceUpdatedAt: null })));
  const toRead = messageIds.filter((id) => !known.has(id));

  // The most useful line this run writes. What follows is one paced request per message --
  // minutes for a real mailbox -- and until now the first sign of how long that would take
  // was the run ending. A total up front turns a blank screen into a quantity, and `skipped`
  // is what keeps "the mailbox is empty" and "nothing has changed" from being one sentence.
  journal.info("work_listed", { entity: ENTITY, total: messageIds.length, skipped: known.size });

  let read = 0;
  for (const messageId of toRead) {
    // The denominator is what is left to do, not what the mailbox holds: on a steady-state
    // mailbox the second is a gauge frozen at zero out of thousands.
    journal.progress("records_read", { entity: ENTITY, read, total: toRead.length });
    const message = await api.getJson(messageUrl(messageId), ENTITY, read);
    const labelIds = strings(getPath(message, "labelIds"));

    // Defence in depth: the query said what to fetch, this says what may be kept. A label
    // removed between the listing and the fetch, or a query built wrong, must not widen
    // what we store beyond what the admin agreed to.
    if (selected.size > 0 && !labelIds.some((id) => selected.has(id))) {
      continue;
    }

    const headers = headerMap(message);
    const internalDate = str(message, "internalDate");
    read += 1;
    yield {
      record: messageRecord(messageId, message, labelIds, internalDate),
      documents: matchingParts(message, scope.fileTypes).map((part) =>
        attachment(api, { messageId, headers, labelIds, internalDate }, part),
      ),
    };
  }

  // Never `seenIds`. A message that stopped matching a label selection was relabelled, not
  // deleted, and a tombstone would report a deletion that never happened.
  return { seenIds: null, skipped: [], listed: messageIds.length, known: known.size };
}

/**
 * One message as a record to land.
 *
 * Built from extracted fields, then canonicalised -- never re-serialised from a parsed
 * payload, and carrying no plain JS number (`canonicalJson` refuses one, and an id that had
 * been through a float would be a different id).
 */
function messageRecord(
  messageId: string,
  message: unknown,
  labelIds: string[],
  internalDate: string,
): RecordToLand {
  return {
    entity: ENTITY,
    sourceRecordId: messageId,
    sourceUpdatedAt: isoFromEpochMillis(internalDate),
    payloadText: canonicalJson({
      id: messageId,
      threadId: str(message, "threadId"),
      labelIds,
      headers: headerMap(message),
      snippetOmitted: true,
      internalDate,
    }),
  };
}

/** What a message says about itself that its attachments need. */
interface MessageFacts {
  readonly messageId: string;
  readonly headers: Record<string, string>;
  readonly labelIds: string[];
  readonly internalDate: string;
}

/**
 * One matching attachment as a document to land.
 *
 * The split between `metadata` and `manifest` is the whole point and it is load-bearing:
 * `metadata` reaches `raw.documents`, which dbt and BI can read, so it carries opaque ids,
 * enumerations and counts only. Every name a human wrote goes in `manifest`, which lives in
 * the access-controlled object store. `pii.md`, ADR 0015.
 */
function attachment(api: GoogleApi, facts: MessageFacts, part: MatchingPart): DocumentToLand {
  const { messageId, headers, labelIds, internalDate } = facts;
  const { attachmentId } = part;

  return {
    // (messageId, partIndex), never attachmentId. See the module docstring.
    documentId: `${messageId}:${String(part.index).padStart(3, "0")}`,
    // An attachment Gmail could only call `application/octet-stream` is filed under what its
    // extension says when the catalogue knows it -- a `.oa` as JSON. `fileFormats.ts`.
    contentType: landedType({ mimeType: part.mimeType, name: part.filename }),
    declaredBytes: part.size,
    metadata: { labelIds, partIndex: String(part.index), messageId },
    manifest: {
      filename: part.filename,
      subject: headers.Subject ?? "",
      from: headers.From ?? "",
      to: headers.To ?? "",
      messageId,
    },
    sourceUpdatedAt: isoFromEpochMillis(internalDate),
    fetchBytes: async (): Promise<Uint8Array> => {
      const body = await api.getJson(
        `${GMAIL_BASE}/messages/${encodeURIComponent(messageId)}/attachments/${encodeURIComponent(attachmentId)}`,
        "attachments",
        0,
      );
      return decodeBase64Url(str(body, "data"));
    },
  };
}

/**
 * One paged query per selected label, unioned.
 *
 * A `Set` rather than a list because a message carrying two selected labels is one message.
 * With nothing selected this is a single unfiltered query -- the whole mailbox, which is a
 * decision the admin recorded rather than an absent one.
 */
async function listMessageIds(api: GoogleApi, scope: GmailScope): Promise<string[]> {
  const queries = scope.labels.length === 0 ? [null] : scope.labels.map((label) => label.id);
  const ids = new Set<string>();

  for (const labelId of queries) {
    await collectLabelIds(api, labelId, ids);
  }

  return [...ids];
}

/** The URL of one page of message ids: the label, if any, and the cursor, if any. */
function listUrl(labelId: string | null, pageToken: string | null): string {
  const url = new URL(`${GMAIL_BASE}/messages`);
  url.searchParams.set("maxResults", PAGE_SIZE);
  if (labelId !== null) {
    url.searchParams.set("labelIds", labelId);
  }
  if (pageToken !== null) {
    url.searchParams.set("pageToken", pageToken);
  }
  return url.toString();
}

/**
 * Every message id under one label, page by page, added to the union.
 *
 * Sequential because each page's cursor comes out of the one before it -- there is nothing to
 * parallelise, and Gmail's pacing is per request either way.
 */
async function collectLabelIds(
  api: GoogleApi,
  labelId: string | null,
  ids: Set<string>,
): Promise<void> {
  let pageToken: string | null = null;
  do {
    const page = await api.getJson(listUrl(labelId, pageToken), ENTITY, ids.size);
    for (const message of asArray(getPath(page, "messages"))) {
      const id = str(message, "id");
      if (id !== "") {
        ids.add(id);
      }
    }
    const next = str(page, "nextPageToken");
    pageToken = next === "" ? null : next;
  } while (pageToken !== null);
}

/**
 * `format=full`, which is the ONLY format that carries attachments.
 *
 * Google's Format reference: `metadata` "Returns only email message ID, labels, and email
 * headers" -- no `payload.parts`, so no `body.attachmentId`, so nothing for `matchingParts`
 * to walk. This asked for `metadata` and therefore landed zero attachments for every tenant,
 * while the suite stayed green because its fixture answered a metadata request with parts.
 *
 * `metadataHeaders` is dropped because it does nothing at this format; `headerMap` keeps the
 * same six headers on the way in instead. What comes back that we do NOT want -- bodies,
 * snippet, the full header set -- is never read and never landed, which is the part that
 * matters and the part `headerMap` and `messageRecord` are pinned on.
 */
function messageUrl(messageId: string): string {
  const url = new URL(`${GMAIL_BASE}/messages/${encodeURIComponent(messageId)}`);
  url.searchParams.set("format", "full");
  return url.toString();
}

interface MatchingPart {
  readonly index: number;
  readonly attachmentId: string;
  readonly filename: string;
  readonly size: string;
  readonly mimeType: string;
}

/**
 * Every attachment in a message whose type the scope allows, walked depth-first.
 *
 * Recursive because a forwarded mail nests `parts` inside `parts`, and an attachment two
 * levels down is still an attachment. The index counts every part visited, so it is stable
 * for a given message shape -- which is what makes it usable as half of the document id.
 */
function matchingParts(message: unknown, fileTypes: readonly string[]): MatchingPart[] {
  const found: MatchingPart[] = [];
  let index = 0;

  function walk(part: unknown): void {
    index += 1;
    const mimeType = str(part, "mimeType");
    const attachmentId = str(part, "body.attachmentId");
    const filename = str(part, "filename");
    if (allowsFile(fileTypes, { mimeType, name: filename }) && attachmentId !== "") {
      found.push({
        index,
        attachmentId,
        filename,
        size: str(part, "body.size") || "0",
        mimeType,
      });
    }
    for (const child of asArray(getPath(part, "parts"))) {
      walk(child);
    }
  }

  for (const part of asArray(getPath(message, "payload.parts"))) {
    walk(part);
  }
  return found;
}

/**
 * The six headers the consent names, keyed by their lowercase spelling.
 *
 * A sender writes the header name and Gmail passes it through, so `Subject` and `SUBJECT`
 * are the same header; matching case-insensitively and storing under the canonical spelling
 * is what keeps `headers.Subject` answering for both.
 */
const KEPT_HEADERS: ReadonlyMap<string, string> = new Map(
  HEADERS.map((name) => [name.toLowerCase(), name] as const),
);

/**
 * The consent's six headers, and nothing else.
 *
 * This filter is load-bearing NOW in a way it was not before. `format=metadata` returned only
 * the headers `metadataHeaders` asked for, so taking all of them took six; `format=full`
 * returns every header a message carries -- Received chains, DKIM signatures, X-* internals,
 * each of which is more about the sender than the invoice is. Widening the fetch to reach
 * attachments must not widen what lands in `raw.records`, so the narrowing moves here.
 */
function headerMap(message: unknown): Record<string, string> {
  const out: Record<string, string> = {};
  for (const header of asArray(getPath(message, "payload.headers"))) {
    const canonical = KEPT_HEADERS.get(str(header, "name").toLowerCase());
    if (canonical !== undefined) {
      out[canonical] = str(header, "value");
    }
  }
  return out;
}

/**
 * Gmail's `internalDate` is epoch milliseconds as a string.
 *
 * Unreadable is `null`, never a guess and never `0` -- an epoch-zero timestamp downstream
 * is indistinguishable from a real 1970 date, and a wrong value is worse than a missing one.
 */
const EPOCH_MILLIS = /^\d+$/u;

function isoFromEpochMillis(value: string): string | null {
  if (!EPOCH_MILLIS.test(value)) {
    return null;
  }
  // parseInt, not Number(): the money rule bans Number() repo-wide, and this is an
  // epoch-millis count rather than an amount. Same reasoning as core's fetcher.
  const date = new Date(Number.parseInt(value, 10));
  return Number.isNaN(date.getTime()) ? null : date.toISOString();
}

function asArray(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

/**
 * A string at a dotted path, or `""`.
 *
 * Through `getStringPath` rather than `String(...)`: a losslessly parsed number is an
 * object, and coercing it is how an exact id becomes "[object Object]". Core already solved
 * this for the connector runtime and the answer is the same here.
 */
function str(root: unknown, path: string): string {
  return getStringPath(root, path) ?? "";
}

/**
 * The string elements of an array, skipping anything else.
 *
 * Gmail sends label ids as strings. Something that is not a string is not an id we can use,
 * and skipping it is honest where coercing it would invent one.
 */
function strings(value: unknown): string[] {
  return asArray(value).filter((v): v is string => typeof v === "string");
}
